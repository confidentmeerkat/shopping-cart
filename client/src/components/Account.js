import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { getUser } from '../actions/userActions';
import AccountModal from './AccountModal';
import Divider from 'material-ui/Divider';
import RaisedButton from 'material-ui/RaisedButton';
import Edit from 'material-ui/svg-icons/image/edit.js';
import '../styles/Account.css';

class Account extends React.Component {

  state = {
    accountModalOpen: false
  }

  toggleAccountModal = () => {
    this.setState({ accountModalOpen: !this.state.accountModalOpen });
  }

  componentDidMount() {
    this.props.getUser();
  }

  render() {
    if(!this.props.user) {
      return (
        <div className="account-container">
          <div className="loader">
            <img src="/img/loader.gif" />
            <h1>LOADING ACCOUNT DATA...</h1>
          </div>
        </div>
      );
    } else {
      return (
        <div className="account-container">
          <h1>Your Account</h1>
          <div className="account">
            <div className="account-info">
              <h2>Info</h2>
              <Divider />
              <p><b>Username: </b>{this.props.user.username}</p>
              <p><b>E-mail: </b>{this.props.user.email}</p>
              <p><b>Billing Address: </b>{this.props.user.address}</p>
              <p><b>Phone: </b>{this.props.user.phone}</p>
              <RaisedButton
                className="btn"
                label="Edit account"
                labelPosition="before"
                primary={true}
                icon={<Edit />}
                onClick={this.toggleAccountModal}
              />
            </div>
            <div className="account-history">
              <h2>Order History</h2>
              <Divider />
              <div class="orders">
                {this.props.user.orders.length ? 
                  <table>
                    <tr>
                      <th>Date Created</th>
                      <th>Product Name</th>
                      <th>Price</th>
                      <th>Qty</th>
                      <th>Total</th>
                    </tr>
                    {this.props.user.orders.map((order) => {
                      return (
                        <tr>
                          <td>{moment(order.dateCreated).format('ll')}</td>
                          <td>{order.name}</td>
                          <td>{numeral(order.price).format('$0,0.00')}</td>
                          <td>{order.quantity}</td>
                          <td>{numeral(order.price * order.quantity).format('$0,0.00')}</td>
                        </tr>
                      );
                    })}
                  </table> :
                  <h1>No order history.</h1>
                }
              </div>
            </div>
          </div>
          <AccountModal
            user={this.props.user}
            isOpen={this.state.accountModalOpen}
            onRequestClose={this.toggleAccountModal}
          />
        </div>
      );
    }
  }
};

const mapStateToProps = (state) => ({
  user: state.loggedUser
});

const mapDispatchToProps = (dispatch) => ({
  getUser: () => dispatch(getUser())
});

export default connect(mapStateToProps, mapDispatchToProps)(Account);