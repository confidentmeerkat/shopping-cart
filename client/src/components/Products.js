import React from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions';
import { clearFilters } from '../actions/filterActions';
import { setSortBy } from '../actions/sortActions';
import filterProducts from '../selectors/filterProducts';
import sortProducts from '../selectors/sortProducts';
import RaisedButton from 'material-ui/RaisedButton';
import Drawer from 'material-ui/Drawer';
import SelectField from 'material-ui/SelectField';
import MenuItem from 'material-ui/MenuItem';
import FiltersList from './FiltersList';
import Product from './Product';
import '../styles/Products.css';

export class Products extends React.Component {
  state = {
    drawerOpen: false,
    value: 'Name: A-Z'
  }

  toggleDrawer = () => {
    this.setState({ drawerOpen: !this.state.drawerOpen });
  }

  handleChange = (event, index, value) => {
    this.props.setSortBy(value);
    this.setState({ value });
  }

  componentWillMount() {
    this.props.initCatalog();
  }
  
  render() {
    if(!this.props.catalogLoaded) {
      return (
        <div className="loader">
          <img src="/img/loader.gif" />
          <h1>LOADING PRODUCTS...</h1>
        </div>
      );
    } else return (
      <div className="products">
        <div className="products-handle">
          <div className="products-found">
            <span><b>Products found: </b>{this.props.catalog.length}</span>
          </div>
          <div className="filters">
            <div className="set-filters">
              <RaisedButton
                className="btn"
                label="Filter products"
                onClick={this.toggleDrawer}
                primary={true}
              />
            </div>
            <RaisedButton
              className="btn"
              label="Clear Filters"
              onClick={this.props.clearFilters}
              secondary={true}
            />
          </div>
          <div className="products-sort">
            <span><b>Sort By:</b></span>
            <SelectField
              className="sort-field"
              value={this.state.value}
              onChange={this.handleChange}
            >
              <MenuItem value="Name: A-Z" primaryText="Name: A-Z" />
              <MenuItem value="Name: Z-A" primaryText="Name: Z-A" />
              <MenuItem value="Price: Low to High" primaryText="Price: Low to High" />
              <MenuItem value="Price: High to Low" primaryText="Price: High to Low" />
            </SelectField>
            <Drawer 
              docked={false}
              width={200}
              open={this.state.drawerOpen}
              onRequestChange={this.toggleDrawer}
            >
              <FiltersList />
            </Drawer>
          </div>
        </div>
        {this.props.catalog.length ? 
          this.props.catalog.map((item) => {
            return <Product key={item.info.name} item={item} />
          }) : 
          <h1 className="no-products">No products found.</h1>}
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  catalogLoaded: state.catalog.isLoaded,
  catalog: sortProducts(filterProducts(state.catalog.items, state.filters), state.sortBy)
});

const mapDispatchToProps = (dispatch) => ({
  initCatalog: () => dispatch(actions.initCatalog()),
  setSortBy: (sortBy) => dispatch(setSortBy(sortBy)),
  clearFilters: () => dispatch(clearFilters())
});

export default connect(mapStateToProps, mapDispatchToProps)(Products);