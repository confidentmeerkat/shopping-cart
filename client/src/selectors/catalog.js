import { createSelector } from 'reselect';

export const isCatalogLoaded = (state) => state.catalog.isLoaded;
export const selectFilters = (state) => state.filters;
export const selectSortBy = (state) => state.sortBy;
export const selectProducts = (state) => state.catalog.items;
export const selectProduct = (state, props) => state.catalog.items.find((item) => item._id == props.match.params.id);

export const sortProducts = (catalog, sortBy) => {
  switch (sortBy) {
    case 'Name: A-Z':
      return catalog.sort((a, b) => (a.info.name > b.info.name) ? 1 : ((b.info.name > a.info.name) ? -1 : 0));
    case 'Name: Z-A':
      return catalog.sort((a, b) => (a.info.name < b.info.name) ? 1 : ((b.info.name < a.info.name) ? -1 : 0));
    case 'Price: Low to High':
      return catalog.sort((a, b) => (a.info.price > b.info.price) ? 1 : ((b.info.price > a.info.price) ? -1 : 0));
    case 'Price: High to Low':
      return catalog.sort((a, b) => (a.info.price < b.info.price) ? 1 : ((b.info.price < a.info.price) ? -1 : 0));
    default:
      return catalog;
  }
};

export const filterProducts = createSelector(
  [selectProducts, selectFilters],
  (catalog, filters) => {
    return catalog.filter((item) => {
      const priceRange = filters.priceRange.length ? filters.priceRange.includes(item.tags.priceRange) : item.tags.priceRange;
      const brand = filters.brand.length ? filters.brand.includes(item.tags.brand) : item.tags.brand;
      const color = filters.color.length ? filters.color.includes(item.tags.color) : item.tags.color;
      const os = filters.os.length ? filters.os.includes(item.tags.os) : item.tags.os;
      const internalMemory = filters.internalMemory.length ? filters.internalMemory.includes(item.tags.internalMemory) : item.tags.internalMemory;
      const ram = filters.ram.length ? filters.ram.includes(item.tags.ram) : item.tags.ram;
      const displaySize = filters.displaySize.length ? filters.displaySize.includes(item.tags.displaySize) : item.tags.displaySize;
      const displayResolution = filters.displayResolution.length ? filters.displayResolution.includes(item.tags.displayResolution) : item.tags.displayResolution;
      const camera = filters.camera.length ? filters.camera.includes(item.tags.camera) : item.tags.camera;
      const cpu = filters.cpu.length ? filters.cpu.includes(item.tags.cpu) : item.tags.cpu;

      return priceRange && brand && color && os && internalMemory && ram && displaySize && displayResolution && camera && cpu;
    })
  }
);
