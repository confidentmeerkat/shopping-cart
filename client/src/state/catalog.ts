import { Product } from './product';

export interface Catalog {
  isLoading: boolean;
  isLoaded: boolean;
  items: Product[];
  error: string | null;
}
