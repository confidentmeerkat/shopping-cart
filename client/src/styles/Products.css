.loader {
  margin: 100px auto;
  color: #00BCD4;
}

.products {
  width: 78%;
  margin: 10px 0;
}

.products-handle {
  display: flex;
  justify-content: space-between;
  padding: 10px;
  border-bottom: 1px solid #81D4FA;
}

.products-handle .left {
  height: 37px;
  margin: 10px 0;
}

.products-handle .left .btn {
  margin-left: 70px;
}

.products-handle .right {
  display: flex;
  justify-content: space-between;
  width: 350px;
  height: 50px;
}

.products-handle .right span {
  margin: 18px 20px 0 0;
}