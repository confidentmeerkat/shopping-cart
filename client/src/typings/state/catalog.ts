import { ICatalogProduct } from './catalogProduct';

export interface ICatalog {
  isLoading: boolean;
  isLoaded: boolean;
  items: ICatalogProduct[];
  error: string | null;
}
