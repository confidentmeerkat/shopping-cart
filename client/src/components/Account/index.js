import { compose } from 'recompose';
import { connect } from 'react-redux';
import { getUser } from '@actions';
import Account from './Account';

const mapStateToProps = (state) => ({
  user: state.loggedUser.user
});

const actions = { getUser };

export default compose(
  connect(mapStateToProps, actions)
)(Account);
