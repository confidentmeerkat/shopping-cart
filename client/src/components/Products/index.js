import { compose } from 'recompose';
import { connect } from 'react-redux';
import { initCatalog, clearFilters, setSortBy } from '@actions';
import { isCatalogLoaded, sortProducts, filterProducts } from '@selectors/catalog';
import Products from './Products';

const mapStateToProps = (state) => ({
  catalogLoaded: isCatalogLoaded(state),
  catalog: sortProducts(filterProducts(state), state.sortBy)
});

const actions = {
  initCatalog,
  clearFilters,
  setSortBy
};

export default compose(
  connect(mapStateToProps, actions)
)(Products);
