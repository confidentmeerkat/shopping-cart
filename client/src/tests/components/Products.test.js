import React from 'react';
import { shallow } from 'enzyme';
import { Products } from '../../components/Products';
import catalog from '../__mocks__/catalog';

describe('<Products />', () => {
  it('renders Products correctly', () => {
    const wrapper = shallow(<Products
      catalogLoaded={true}
      catalog={catalog}
      initCatalog={() => { }}
      setSortBy={() => { }}
      clearFilters={() => { }}
    />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Product')).toHaveLength(3);
  });

  it('sorts products by given criteria', () => {
    const wrapper = shallow(<Products
      catalogLoaded={true}
      catalog={catalog}
      initCatalog={() => { }}
      setSortBy={() => { }}
      clearFilters={() => { }}
    />);

    wrapper.find('SelectField').simulate('change', null, null, 'Price: Low to High');
    expect(wrapper.state('value')).toBe('Price: Low to High');
  });
});