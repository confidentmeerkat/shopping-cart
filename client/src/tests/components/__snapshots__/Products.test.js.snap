// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> renders Products correctly 1`] = `
<div
  className="products"
>
  <div
    className="products-handle"
  >
    <div
      className="products-found"
    >
      <span>
        <b>
          Products found: 
        </b>
        3
      </span>
    </div>
    <div
      className="filters"
    >
      <div
        className="set-filters"
      >
        <RaisedButton
          className="btn"
          disabled={false}
          fullWidth={false}
          label="Filter products"
          labelPosition="after"
          onClick={[Function]}
          primary={true}
          secondary={false}
        />
      </div>
      <RaisedButton
        className="btn"
        disabled={false}
        fullWidth={false}
        label="Clear Filters"
        labelPosition="after"
        onClick={[Function]}
        primary={false}
        secondary={true}
      />
    </div>
    <div
      className="products-sort"
    >
      <span>
        <b>
          Sort By:
        </b>
      </span>
      <SelectField
        autoWidth={false}
        className="sort-field"
        disabled={false}
        fullWidth={false}
        multiple={false}
        onChange={[Function]}
        value="Name: A-Z"
      >
        <MenuItem
          anchorOrigin={
            Object {
              "horizontal": "right",
              "vertical": "top",
            }
          }
          checked={false}
          desktop={false}
          disabled={false}
          focusState="none"
          insetChildren={false}
          primaryText="Name: A-Z"
          targetOrigin={
            Object {
              "horizontal": "left",
              "vertical": "top",
            }
          }
          value="Name: A-Z"
        />
        <MenuItem
          anchorOrigin={
            Object {
              "horizontal": "right",
              "vertical": "top",
            }
          }
          checked={false}
          desktop={false}
          disabled={false}
          focusState="none"
          insetChildren={false}
          primaryText="Name: Z-A"
          targetOrigin={
            Object {
              "horizontal": "left",
              "vertical": "top",
            }
          }
          value="Name: Z-A"
        />
        <MenuItem
          anchorOrigin={
            Object {
              "horizontal": "right",
              "vertical": "top",
            }
          }
          checked={false}
          desktop={false}
          disabled={false}
          focusState="none"
          insetChildren={false}
          primaryText="Price: Low to High"
          targetOrigin={
            Object {
              "horizontal": "left",
              "vertical": "top",
            }
          }
          value="Price: Low to High"
        />
        <MenuItem
          anchorOrigin={
            Object {
              "horizontal": "right",
              "vertical": "top",
            }
          }
          checked={false}
          desktop={false}
          disabled={false}
          focusState="none"
          insetChildren={false}
          primaryText="Price: High to Low"
          targetOrigin={
            Object {
              "horizontal": "left",
              "vertical": "top",
            }
          }
          value="Price: High to Low"
        />
      </SelectField>
      <Drawer
        disableSwipeToOpen={false}
        docked={false}
        onRequestChange={[Function]}
        open={false}
        openSecondary={false}
        swipeAreaWidth={30}
        width={200}
        zDepth={2}
      >
        <Connect(FiltersList) />
      </Drawer>
    </div>
  </div>
  <Product
    item={
      Object {
        "info": Object {
          "batery": "Non-removable Li-Ion 2691 mAh battery (10.28 Wh)",
          "camera": "Dual: 12 MP (f/1.8, 28mm, OIS) + 12 MP (f/2.8, 57mm)",
          "color": "White",
          "cpu": "Hexa-core (2x Monsoon + 4x Mistral)",
          "dimensions": "158.4 x 78.1 x 7.5 mm",
          "displayResolution": "1080 x 1920 pixels",
          "displaySize": "5.5\\"",
          "displayType": "LED-backlit IPS LCD, capacitive touchscreen, 16M colors",
          "internalMemory": "256 GB",
          "name": "Apple iPhone 8 Plus",
          "os": "iOS 11",
          "photo": "/img/apple_iphone_8_plus.jpg",
          "price": 700,
          "ram": "3 GB",
          "weight": "202 g",
        },
      }
    }
    key="Apple iPhone 8 Plus"
  />
  <Product
    item={
      Object {
        "info": Object {
          "batery": "Non-removable Li-Po 3300 mAh battery",
          "camera": "Dual: 13 MP (f/1.8, 1/3\\", 1.12 µm, 3-axis OIS, PDAF) + 13 MP(f / 2.4, no AF)",
          "color": "Ice Platinum",
          "cpu": "Quad-core (2x2.35 GHz Kryo & 2x1.6 GHz Kryo)",
          "dimensions": "148.9 x 71.9 x 7.9 mm",
          "displayResolution": "1440 x 2880 pixels",
          "displaySize": "5.8\\"",
          "displayType": "IPS LCD capacitive touchscreen, 16M colors",
          "internalMemory": "128 GB",
          "name": "LG G6",
          "os": "Android 7.0 (Nougat)",
          "photo": "/img/lg_g6.jpg",
          "price": 800,
          "ram": "4 GB",
          "weight": "163 g",
        },
      }
    }
    key="LG G6"
  />
  <Product
    item={
      Object {
        "info": Object {
          "batery": "Non-removable Li-Ion 3000 mAh battery",
          "camera": "12 MP (f/1.7, 26mm, 1/2.5\\", 1.4 µm, Dual Pixel PDAF), phase detection autofocus, OIS",
          "color": "Midnight Black",
          "cpu": "Octa-core (4x2.3 GHz & 4x1.7 GHz) - EMEA",
          "dimensions": "148.9 x 68.1 x 8 mm",
          "displayResolution": "1440 x 2960 pixels",
          "displaySize": "5.8\\"",
          "displayType": "Super AMOLED capacitive touchscreen, 16M colors",
          "internalMemory": "64 GB",
          "name": "Samsung Galaxy S8",
          "os": "Android 7.0 (Nougat)",
          "photo": "/img/samsung_galaxy_s8.jpg",
          "price": 720,
          "ram": "4 GB",
          "weight": "155 g",
        },
      }
    }
    key="Samsung Galaxy S8"
  />
</div>
`;
